### PR TITLE
LPS 83417 - Add MyUserAccount API

### PR DESCRIPTION
**Pull request by eperezg**

### CODE REVIEW

File: modules/apps/headless-apio/person/person-apio-impl/bnd.bnd [hunk 1]
```diff
@@ -1,3 +1,4 @@
 Bundle-Name: Liferay Person Apio Implementation
 Bundle-SymbolicName: com.liferay.person.apio.impl
-Bundle-Version: 2.0.0
\ No newline at end of file
+Bundle-Version: 2.0.0
+-dsannotations-options: inherit
```
Comment (nhpatt) [inline modules/apps/headless-apio/person/person-apio-impl/bnd.bnd]: mmm... we should talk about this, we have followed another approach for reusing code in other APIs

File: modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/base/BaseUserAccountCollectionResource.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	protected void addCustomFields(
+		Representor.FirstStep<UserWrapper> builderFirstStep) {
+	}
+
+	protected UserWrapper addUser(
+		PersonCreatorForm personCreatorForm, ThemeDisplay themeDisplay) {
+
+		return null;
```
Comment (nhpatt) [inline modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/base/BaseUserAccountCollectionResource.java]: this is not very pretty :P

File: modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/base/BaseUserAccountCollectionResource.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	protected abstract PageItems<UserWrapper> getPageItems(
+			Pagination pagination, FullNameQuery fullNameQuery,
+			Credentials credentials, ThemeDisplay themeDisplay,
+			CurrentUser currentUser)
+		throws PortalException;
+
+	protected UserWrapper updateUser(
```
Comment (nhpatt) [inline modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/base/BaseUserAccountCollectionResource.java]: why?

Comment (nhpatt) [reply modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/base/BaseUserAccountCollectionResource.java]: Can you return the builder in the base class and add the update/create part in the child classes?

File: modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/MyUserAccountCollectionResource.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.Arrays;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * Provides the information necessary to expose MyUserAccount resources through
+ * a web API. The resources are mapped from the internal model
+ * {@link UserWrapper}.
+ *
+ * @author Eduardo Perez
```
Comment (nhpatt) [inline modules/apps/headless-apio/person/person-apio-impl/src/main/java/com/liferay/person/apio/internal/architect/resource/MyUserAccountCollectionResource.java]: @review

File: modules/apps/headless-apio/organization/organization-apio-impl/src/main/java/com/liferay/organization/apio/internal/architect/router/MyUserAccountUserAccountWebSiteNestedCollectionRouter.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ * Provides the information necessary to expose the <a
+ * href="http://schema.org/Organization">Organization</a> resources contained
+ * inside a MyUserAccount through a web API. The resources are mapped from the
+ * internal model {@link Organization}.
+ *
+ * @author Eduardo Perez
+ * @review
+ */
+@Component(immediate = true)
+public class MyUserAccountUserAccountWebSiteNestedCollectionRouter extends
```
Comment (nhpatt) [inline modules/apps/headless-apio/organization/organization-apio-impl/src/main/java/com/liferay/organization/apio/internal/architect/router/MyUserAccountUserAccountWebSiteNestedCollectionRouter.java]: the name is really strange

File: modules/apps/headless-apio/organization/organization-apio-impl/src/main/java/com/liferay/organization/apio/internal/architect/router/PersonUserAccountWebSiteNestedCollectionRouter.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ * Provides the information necessary to expose the <a
+ * href="http://schema.org/Organization">Organization</a> resources contained inside a <a
+ * href="http://schema.org/Person">Person</a> through a web API. The resources
+ * are mapped from the internal model {@link Organization}.
+ *
+ * @author Eduardo Perez
+ * @review
+ */
+@Component(immediate = true)
+public class PersonUserAccountWebSiteNestedCollectionRouter extends
```
Comment (nhpatt) [inline modules/apps/headless-apio/organization/organization-apio-impl/src/main/java/com/liferay/organization/apio/internal/architect/router/PersonUserAccountWebSiteNestedCollectionRouter.java]: Name

File: modules/apps/headless-apio/address/address-apio-impl/src/main/java/com/liferay/address/apio/internal/architect/router/base/BaseUserAccountNestedCollectionRouter.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+		User user = userService.getUserById(personId);
+
+		String className = user.getModelClassName();
+
+		CommonPermissionUtil.check(
+			(PermissionChecker)credentials.get(), className, user.getUserId(),
+			ActionKeys.VIEW);
+
+		List<Address> addresses = addressLocalService.getAddresses(
```
Comment (nhpatt) [inline modules/apps/headless-apio/address/address-apio-impl/src/main/java/com/liferay/address/apio/internal/architect/router/base/BaseUserAccountNestedCollectionRouter.java]: Do the Math.min thing

File: modules/apps/headless-apio/email/email-apio-impl/src/main/java/com/liferay/email/apio/internal/architect/router/base/BaseUserAccountEmailsNestedCollectionRouter.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	@Reference
+	protected UserService userService;
+
+	private PageItems<EmailAddress> _getPageItems(
+			Pagination pagination, long personId)
+		throws PortalException {
+
+		User user = userService.getUserById(personId);
+
+		List<EmailAddress> emailAddresses =
```
Comment (nhpatt) [inline modules/apps/headless-apio/email/email-apio-impl/src/main/java/com/liferay/email/apio/internal/architect/router/base/BaseUserAccountEmailsNestedCollectionRouter.java]: Do the Math.min thing

File: modules/apps/headless-apio/phone/phone-apio-impl/src/main/java/com/liferay/phone/apio/internal/architect/router/base/BaseUserAccountPhonesNestedCollectionRouter.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+	@Reference
+	protected UserService userService;
+
+	private PageItems<Phone> _getPageItems(Pagination pagination, long personId)
+		throws PortalException {
+
+		User user = userService.getUserById(personId);
+
+		List<Phone> phones = phoneService.getPhones(
```
Comment (nhpatt) [inline modules/apps/headless-apio/phone/phone-apio-impl/src/main/java/com/liferay/phone/apio/internal/architect/router/base/BaseUserAccountPhonesNestedCollectionRouter.java]: Do the Math.min thing

File: modules/apps/headless-apio/role/role-apio-impl/src/main/java/com/liferay/role/apio/internal/router/base/BaseUserAccountRoleNestedCollectionRouter.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		).build();
+	}
+
+	@Reference
+	protected RoleService roleService;
+
+	private PageItems<Role> _getPageItems(Pagination pagination, long userId)
+		throws PortalException {
+
+		List<Role> roles = roleService.getUserRoles(userId);
```
Comment (nhpatt) [inline modules/apps/headless-apio/role/role-apio-impl/src/main/java/com/liferay/role/apio/internal/router/base/BaseUserAccountRoleNestedCollectionRouter.java]: Do the Math.min thing

File: modules/apps/headless-apio/workflow/workflow-apio-impl/src/main/java/com/liferay/workflow/apio/internal/architect/router/base/BaseWorkflowTasksUserNestedCollectionRouter.java [hunk 11]
```diff
@@ -27,21 +27,15 @@
...
- * WorkflowTask}.
- *
  * @author Sarai Díaz
  * @review
  */
-@Component(immediate = true)
-public class WorkflowTasksUserNestedCollectionRouter
-	implements NestedCollectionRouter
-		<WorkflowTask, Long, WorkflowTaskIdentifier, Long, PersonIdentifier> {
+public abstract class BaseWorkflowTasksUserNestedCollectionRouter
```
Comment (nhpatt) [inline modules/apps/headless-apio/workflow/workflow-apio-impl/src/main/java/com/liferay/workflow/apio/internal/architect/router/base/BaseWorkflowTasksUserNestedCollectionRouter.java]: Name is not consistent, right?

**Comment by eperezg** [comment]
@nhpatt can you take another look?
Thnx

**Pull request merged**